### ISSUE TITLE
results

### ISSUE THREAD

**Issue by XYZOK000** [zhouhaoyi/Informer2020#26]
运行程序后results文件夹中的trues.npy是三维数据，和原二维数据什么关系呢，还有预测结果preds.npy也是三维的，为什么不是像预测目标‘OT’那样是一列呢，不太理解，请多多指教。

**Comment by JYOK00** [comment]
我看最后的mape为4.7，请问为什么相对误差这么大呢

**Comment by cookieminions** [comment]
关于preds.npy的shape问题，可以参考代码`/exp/exp_informer.py`中的第225-226行，其shape应该是`[samples, pred_len, dimensions]`，即每一个sample都是一次预测的结果，其shape是`[pred_len, dimension]`，而trues.npy的shape是与preds.npy对应的

**Comment by cookieminions** [comment]
关于mape的问题，如果mape过大可能与某些个别数据点有关，此外为避免数据标准化对mape的影响，也可以使用经过标准化的输入x和原始大小的数据y，在代码`/exp/exp_informer.py`中的train部分将模型的output还原为原始大小的数值和y进行loss运算

**Comment by JYOK00** [comment]
还有好多问题不太清楚  有什么办法可以私信您吗

**Comment by cookieminions** [comment]
可以将问题整理通过邮件发送给我们

**Comment by JYOK00** [comment]
好的

**Comment by JYOK00** [comment]
之前发过一次 ，我汇总一下

**Comment by JYOK00** [comment]
好多不太会，真的拜托您了。
1.请问output 代码怎么改才可以是原始大小呢
2.我看您的数据是6个变量预测1个吧，换成我的数据的话需要在哪些地方改呢，只要数字是7的地方都要改吗
3.输出的数据shape是pred_len,dimension 预测结果的dimension为什么不是1呢
4.我想像回归那样，划分训练集，验证集，测试集，然后得到一列预测结果，不知道能不能做到呢
求解答

---Original---

**Comment by cookieminions** [comment]
1. 不太明白你说的原始大小是什么含义，如果说的是数值大小，可以通过data_loader的scaler将其inverse_transform回去，这部分可以参考`data/data_loader.py`中的代码，以及sklearn的StandardScaler
2. 目前的多变量预测是7个变量的输入预测7个变量的输出，如果需要修改可以阅读代码`data/data_loader.py`，`exp/exp_informer.py`和`main_informer.py`并自行修改其中的数据加载、模型数据输入以及参数设置等部分；
3. 输出数据的dimension可以自己定义，目前多变量的输出默认是和输入的维度相同，因此如果输入是7维，输出也是7维；
4. 时序预测的训练集、验证集和测试集划分不能是随机打乱划分，而应该按区域切分序列，然后再对切分后的序列取一定长度的子序列作为输入，这部分请参考`data/data_loader.py`。

**Comment by JYOK00** [comment]
那如何设置6个输入，1个输出呢，不太清楚是代码的哪些部分




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zhouhaoyi/Informer2020"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月19日(星期五) 下午2:33
收件人:&nbsp;"zhouhaoyi/Informer2020"<Informer2020@noreply.github.com&gt;;
抄送:&nbsp;"OK"<1175256413@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [zhouhaoyi/Informer2020] results (#26)





  
不太明白你说的原始大小是什么含义，如果说的是数值大小，可以通过data_loader的scaler将其inverse_transform回去，这部分可以参考data/data_loader.py中的代码，以及sklearn的StandardScaler
 
目前的多变量预测是7个变量的输入预测7个变量的输出，如果需要修改可以阅读代码data/data_loader.py，exp/exp_informer.py和main_informer.py并自行修改其中的数据加载、模型数据输入以及参数设置等部分；
 
输出数据的dimension可以自己定义，目前多变量的输出默认是和输入的维度相同，因此如果输入是7维，输出也是7维；
 
时序预测的训练集、验证集和测试集划分不能是随机打乱划分，而应该按区域切分序列，然后再对切分后的序列取一定长度的子序列作为输入，这部分请参考data/data_loader.py。

**Comment by cookieminions** [comment]
你可以尝试修改`exp/exp_informer.py`中的第159行，将其改为`batch_y = batch_y[:,-self.args.pred_len:,-1:].to(self.device)`，同时注意修改vali和test方法中的相同部分，这些的目的是使模型输出维度为1时与其计算loss的y维度也为1；此外尝试修改`main_informer.py`中的第49-53行，修改为类似`'ETTh1':{'data':'ETTh1.csv','T':'OT','M':[7,7,1],'S':[1,1,1]},`的代码，如果你阅读了`main_informer.py`的代码，应该是能够发现这部分是一个简化参数赋值的parser，其中M:[7,7,1]对应的是模型的encoder和decoder输入以及输出
以上只是提供一个参考，不保证绝对运行正确，请阅读代码理解实验的运作流程

**Comment by XYZOK000** [comment]
谢谢

**Comment by cookieminions** [comment]
不客气，欢迎讨论

**Comment by XYZOK000** [comment]
这里的pred_len应该是24吧
train 8521
val 2857
test 2857
为什么预测长度不是测试集test的长度2857呢

**Comment by XYZOK000** [comment]
我之后有问题会再给您发邮件的，谢谢您

**Comment by XYZOK000** [comment]
每一个sample都是预测的结果，那以哪一次为准计算mape这些指标呢，是每一次求平均值吗，pred_len和dimensions是指什么呢，和数据集的关系是什么，哪一个参数是用来控制训练集和测试集的长度的呢，我觉得预测结果和测试集的长度对不上。
求解答。
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zhouhaoyi/Informer2020"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月19日(星期五) 中午12:02
收件人:&nbsp;"zhouhaoyi/Informer2020"<Informer2020@noreply.github.com&gt;;
抄送:&nbsp;"OK"<1175256413@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [zhouhaoyi/Informer2020] results (#26)





 
关于preds.npy的shape问题，可以参考代码/exp/exp_informer.py中的第225-226行，其shape应该是[samples, pred_len, dimensions]，即每一个sample都是一次预测的结果，其shape是[pred_len, dimension]，而trues.npy的shape是与preds.npy对应的

**Comment by cookieminions** [comment]
1. 关于mape：所有sample求平均，计算mape的过程在`exp/exp_informer.py`中存在，位于第245行，请仔细阅读代码
2. 关于pred_len和dimension：关于pred_len指预测长度，dimension指输出的维度即c_out，请仔细阅读`main_informer.py`的代码
3. 输出的测试集长度为test_data的长度，并不是test_loader的长度，因为在建立dataloader时使用了drop last，会丢弃掉最后组不成足够batch_size数量的那几个sample，将`exp/exp_informer.py`中第68行代码位置的test_loader的batch_size修改为1，则会得到测试集和预测结果一样的sample数了

最后请仔细阅读代码并进行实验

**Comment by XYZOK000** [comment]
太打扰您啦，这些我知道在代码中都有，就是我不知道它对应的含义才总是问您，真的很抱歉。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zhouhaoyi/Informer2020"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月19日(星期五) 晚上10:55
收件人:&nbsp;"zhouhaoyi/Informer2020"<Informer2020@noreply.github.com&gt;;
抄送:&nbsp;"OK"<1175256413@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [zhouhaoyi/Informer2020] results (#26)





  
关于mape：所有sample求平均，计算mape的过程在exp/exp_informer.py中存在，位于第245行，请仔细阅读代码
 
关于pred_len和dimension：关于pred_len指预测长度，dimension指输出的维度即c_out，请仔细阅读main_informer.py的代码
 
输出的测试集长度为test_data的长度，并不是test_loader的长度，因为在建立dataloader时使用了drop last，会丢弃掉最后组不成足够batch_size数量的那几个sample，将exp/exp_informer.py中第68行代码位置的test_loader的batch_size修改为1，则会得到测试集和预测结果一样的sample数了
  
最后请仔细阅读代码并进行实验

**Comment by XYZOK000** [comment]
您好，我还有问题想问。
exp/exp_informer.py中true = batch_y.detach().cpu().numpy()#.squeeze()
如果没有标准化的话，true是不是就是原始数据的真实值，但我的数据中没有0，但打印出来看却有0，true又是什么的结果呢


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zhouhaoyi/Informer2020"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月19日(星期五) 晚上10:55
收件人:&nbsp;"zhouhaoyi/Informer2020"<Informer2020@noreply.github.com&gt;;
抄送:&nbsp;"OK"<1175256413@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [zhouhaoyi/Informer2020] results (#26)





  
关于mape：所有sample求平均，计算mape的过程在exp/exp_informer.py中存在，位于第245行，请仔细阅读代码
 
关于pred_len和dimension：关于pred_len指预测长度，dimension指输出的维度即c_out，请仔细阅读main_informer.py的代码
 
输出的测试集长度为test_data的长度，并不是test_loader的长度，因为在建立dataloader时使用了drop last，会丢弃掉最后组不成足够batch_size数量的那几个sample，将exp/exp_informer.py中第68行代码位置的test_loader的batch_size修改为1，则会得到测试集和预测结果一样的sample数了
  
最后请仔细阅读代码并进行实验

**Comment by XYZOK000** [comment]
这个我研究明白了，打扰了
您好，我还有问题想问。
exp/exp_informer.py中true = batch_y.detach().cpu().numpy()#.squeeze()
如果没有标准化的话，true是不是就是原始数据的真实值，但我的数据中没有0，但打印出来看却有0，true又是什么的结果呢


------------------ 原始邮件 ------------------
发件人:                                                                                                                        "zhouhaoyi/Informer2020"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年2月19日(星期五) 晚上10:55
收件人:&nbsp;"zhouhaoyi/Informer2020"<Informer2020@noreply.github.com&gt;;
抄送:&nbsp;"OK"<1175256413@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [zhouhaoyi/Informer2020] results (#26)





  
关于mape：所有sample求平均，计算mape的过程在exp/exp_informer.py中存在，位于第245行，请仔细阅读代码
 
关于pred_len和dimension：关于pred_len指预测长度，dimension指输出的维度即c_out，请仔细阅读main_informer.py的代码
 
输出的测试集长度为test_data的长度，并不是test_loader的长度，因为在建立dataloader时使用了drop last，会丢弃掉最后组不成足够batch_size数量的那几个sample，将exp/exp_informer.py中第68行代码位置的test_loader的batch_size修改为1，则会得到测试集和预测结果一样的sample数了
  
最后请仔细阅读代码并进行实验
